tests/test_cli_compile.py

```diff
@@ -612,38 +630,31 @@ def test_local_url_package(
     assert dependency in out.stderr
 
 
-@pytest.mark.parametrize(
-    ("line", "dependency", "rewritten_line"),
-    (
-        pytest.param(
-            os.path.join(
-                MINIMAL_WHEELS_PATH, "small_fake_with_deps-0.1-py2.py3-none-any.whl"
-            ),
-            "\nfile:small_fake_with_deps-0.1-py2.py3-none-any.whl",
-            "file:small_fake_with_deps-0.1-py2.py3-none-any.whl",
-            id="Relative URL",
+def test_relative_file_uri_package(pip_conf, runner):
+    # Copy wheel into temp dir
+    shutil.copy(
+        os.path.join(
+            MINIMAL_WHEELS_PATH, "small_fake_with_deps-0.1-py2.py3-none-any.whl"
         ),
-        pytest.param(
-            os.path.join(
-                MINIMAL_WHEELS_PATH, "small_fake_with_deps-0.1-py2.py3-none-any.whl"
-            ),
-            "\nsmall-fake-with-deps"
-            " @ file://{absolute_path}/small_fake_with_deps-0.1-py2.py3-none-any.whl",
-            "small-fake-with-deps"
-            " @ file://{absolute_path}/small_fake_with_deps-0.1-py2.py3-none-any.whl",
-            id="Relative URL",
-        ),
-    ),
-)
-def test_relative_url_package(pip_conf, runner, line, dependency, rewritten_line):
-    dependency = dependency.format(absolute_path=Path(".").absolute())
-    rewritten_line = rewritten_line.format(absolute_path=Path(".").absolute())
-    shutil.copy(line, ".")
+        ".",
+    )
     with open("requirements.in", "w") as req_in:
-        req_in.write(dependency)
+        req_in.write("file:small_fake_with_deps-0.1-py2.py3-none-any.whl")
     out = runner.invoke(cli, ["-n", "--rebuild"])
     assert out.exit_code == 0
-    assert rewritten_line in out.stderr
+    assert "file:small_fake_with_deps-0.1-py2.py3-none-any.whl" in out.stderr
+
+
+def test_direct_reference_with_extras(runner):
+    with open("requirements.in", "w") as req_in:
+        req_in.write(
+            "piptools[testing,coverage] @ git+https://github.com/jazzband/pip-tools@6.2.0"
+        )
+    out = runner.invoke(cli, ["-n", "--rebuild"])
+    assert out.exit_code == 0
+    assert "pip-tools @ git+https://github.com/jazzband/pip-tools@6.2.0" in out.stderr
```

Comment:
Shouldn't there be extras included in the package name?
```suggestion
    assert "pip-tools[testing,coverage] @ git+https://github.com/jazzband/pip-tools@6.2.0" in out.stderr
```

Reply:
No, basically at some point it takes into account the extras that is passed in the `.in` like so:

```
❯ cat requirements.in
uvicorn[standard] @ file:///Users/florent/Desktop/fastapi/uvicorn-0.14.0-py3-none-any.whl
```

And so it produces the `.txt` with the extras:
```
❯ cat requirements.txt
#
# This file is autogenerated by pip-compile with python 3.9
# To update, run:
#
#    pip-compile
#
asgiref==3.4.1
    # via uvicorn
click==8.0.1
    # via uvicorn
h11==0.12.0
    # via uvicorn
httptools==0.2.0
    # via uvicorn
python-dotenv==0.18.0
    # via uvicorn
pyyaml==5.4.1
    # via uvicorn
uvicorn @ file:///Users/florent/Desktop/fastapi/uvicorn-0.14.0-py3-none-any.whl
    # via -r requirements.in
uvloop==0.15.3
    # via uvicorn
watchgod==0.7
    # via uvicorn
websockets==9.1
    # via uvicorn
```

If I remove the extras from the `.in`, it gives this:

```
❯ pip-compile
#
# This file is autogenerated by pip-compile with python 3.9
# To update, run:
#
#    pip-compile
#
asgiref==3.4.1
    # via uvicorn
click==8.0.1
    # via uvicorn
h11==0.12.0
    # via uvicorn
uvicorn @ file:///Users/florent/Desktop/fastapi/uvicorn-0.14.0-py3-none-any.whl
    # via -r requirements.in
```

Reply:
Another example is the ouput of `poetry`:

from:
```
pip-tools = {git = "https://github.com/FlorentJeannot/pip-tools.git", rev = "master", extras=["testing"]}
```

the `requirements.txt` contains:
```
pip-tools @ git+https://github.com/FlorentJeannot/pip-tools.git@master ; python_version >= "3.6"
```

not:
```
pip-tools[testing] @ git+https://github.com/FlorentJeannot/pip-tools.git@master ; python_version >= "3.6"
```

So the current behavior of `pip-tools` seems fine to me. 🤔

Reply:
And from `pip`:

```
❯ pip install 'git+https://github.com/encode/uvicorn.git@0.14.0#egg=uvicorn[standard]'
```

`pip freeze`:

```
asgiref==3.4.1
click==8.0.1
h11==0.12.0
httptools==0.2.0
python-dotenv==0.18.0
PyYAML==5.4.1
uvicorn @ git+https://github.com/encode/uvicorn.git@87da6cf4082cd28306bdb78d7d42552d131cc179
uvloop==0.15.3
watchgod==0.7
websockets==9.1
```

So unless the specification was not implemented correctly in the other tools, I think we're fine?

Reply:
Also if we specify the extras in the `requirements.txt` and we have the extras of a package listed in the file as well, I think it's redundant?

Like for `uvicorn` if you add [standard] it's going to install `uvloop` and a bunch of other packages.

So if you have:

```
uvicorn[standard] @ git+https://github.com/encode/uvicorn.git@87da6cf4082cd28306bdb78d7d42552d131cc179
uvloop==0.15.3
```

I think it would mean when it installs `uvicorn`, then it also installs `uvloop`. And then when it wants to install `uvloop` I think it's going to say it's already installed.

Reply:
Thanks for the analysis! My point is that `pip-compile` should be consistent with the following example:

```console
$ echo pip-tools[coverage] | pip-compile - -qo-
...
pip-tools[coverage]==6.2.0
...
```

While stripping extras in `requirements.txt` seems logical, but I'd like to be sure that it doesn't break any user's workflow.

Reply:
Ok I see. I looked at https://github.com/jazzband/pip-tools/issues/154 and I think IMO this was wrong to suggest `hypothesis[django]==3.6.0` as the expected result in `requirements.txt`. There was really an issue though, as it was missing the dependencies of the extras in the actual result.

I think it should have been either this:

```
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --output-file reqs.txt reqs.in
#
click==6.6                # via pip-tools
first==2.0.1              # via pip-tools
hypothesis[django]==3.6.0
pip-tools==1.7.0
six==1.10.0               # via pip-tools
```

or this:

```
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --output-file reqs.txt reqs.in
#
click==6.6                # via pip-tools
django==1.10.3            # via hypothesis
fake-factory==0.7.2       # via hypothesis
first==2.0.1              # via pip-tools
hypothesis==3.6.0
pip-tools==1.7.0
python-dateutil==2.6.0    # via fake-factory
pytz==2016.7              # via hypothesis
six==1.10.0               # via fake-factory, pip-tools, python-dateutil
```

but not:

```
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --output-file reqs.txt reqs.in
#
click==6.6                # via pip-tools
django==1.10.3            # via hypothesis
fake-factory==0.7.2       # via hypothesis
first==2.0.1              # via pip-tools
hypothesis[django]==3.6.0
pip-tools==1.7.0
python-dateutil==2.6.0    # via fake-factory
pytz==2016.7              # via hypothesis
six==1.10.0               # via fake-factory, pip-tools, python-dateutil
```

Regarding the direct reference, it should not break any user's workflow as the direct reference was not supported in the `requirements.in` file. It can be disturbing though, for example we could have users saying the direct reference is missing the extras in the `requirements.txt` like @graingert.

What I suggest is:

1. I add the extras for the direct reference in the `requirements.txt` even though I don't think it makes sense to do that
2. I provide another PR to remove this extras in `requirements.txt` for `pip-tools[coverage]==6.2.0` and direct reference

Reply:
does pip keep track of what extras were requested? eg I install `coverage[toml]` which installs `coverage` and `toml` then sometime later upgrade `coverage` to a version that uses `tomli` will pip know to install `tomli`?

Reply:
Well I think the workflow is this:

I specify a version of coverage with the extras in the `requirements.in`:

```
coverage[toml] @ git+https://github.com/nedbat/coveragepy.git@coverage-5.1
```

Which gives this output:

```
❯ pip-compile
#
# This file is autogenerated by pip-compile with python 3.9
# To update, run:
#
#    pip-compile
#
coverage @ git+https://github.com/nedbat/coveragepy.git@coverage-5.1
    # via -r requirements.in
toml==0.10.2
    # via coverage
```

If I update `coverage` in my `requirements.in` to another version (or branch):

```
coverage[toml] @ git+https://github.com/nedbat/coveragepy.git@master
```

then I must `pip-compile` again, which gives this output:

```
❯ pip-compile
#
# This file is autogenerated by pip-compile with python 3.9
# To update, run:
#
#    pip-compile
#
coverage @ git+https://github.com/nedbat/coveragepy.git@master
    # via -r requirements.in
tomli==1.0.4
    # via coverage
```

and finally when I do `pip-sync` it's going to clean the venv (remove `toml`) and install the right dependencies (install `tomli`).

@graingert Does it make sense or did I miss something?

Reply:
In case you have specified `coverage[toml]` in `requirements.in` and you have already generated the `requirements.txt`(which outputted `toml` as dependency), I think as long as you don't do `pip-compile -U`, it's going to keep the version that has been pinned in the `requirements.txt`.

Reply:
Oh sure, but I'm talking about a mixed workflow. Does pip actually store the extras you select anywhere?

Reply:
I am not sure to understand your workflow so I'll try to answer as best I can.

If I install coverage with pip it does the following:

```
pip install 'coverage[toml] @ git+https://github.com/nedbat/coveragepy.git@coverage-5.1'
[...]

❯ pip freeze
coverage @ git+https://github.com/nedbat/coveragepy.git@b854e6103efd49a2c72438db55e99273ef5fc52e
toml==0.10.2
```

Then if I upgrade:

```
pip install -U 'coverage[toml] @ git+https://github.com/nedbat/coveragepy.git@master'
[...]

❯ pip freeze
coverage @ git+https://github.com/nedbat/coveragepy.git@c0da97eb03d4ffe8be8854ad6ff1a2736f169003
toml==0.10.2
tomli==1.0.4
```

(By the way you can see that `pip` does not display the extras for `coverage` which makes me think we should not do this in the `requirements.txt` when we run `pip-compile`).

Please let me know if I answered your question, otherwise could you give me more details about your workflow?
Thanks!

Reply:
Yep, that's what I feared. The fact that pip-compile left the extras in made me think they were important

Reply:
> Yep, that's what I feared. The fact that pip-compile left the extras in made me think they were important

Yeah I think `pip-compile` is doing it wrong actually. `pip`, `pipenv` and `poetry` are not showing the extras in the `.txt` file when you generate it. I really think we should fix this in `pip-tools` and do it like the other tools.